llarp/router/router.cpp

```diff
@@ -227,14 +229,97 @@ namespace llarp
       LogError(rcfile, " contains invalid RC");
   }
 
+  static size_t
+  RecvIdentKey(char *ptr, size_t, size_t nmemb, void *userdata)
+  {
+    for(size_t idx = 0; idx < nmemb; idx++)
+      static_cast< std::vector< char > * >(userdata)->push_back(ptr[idx]);
+    return nmemb;
+  }
+
   bool
   Router::EnsureIdentity()
   {
     if(!EnsureEncryptionKey())
       return false;
-    if(usingSNSeed)
-      return llarp_loadServiceNodeIdentityKey(ident_keyfile, _identity);
-
+    if(whitelistRouters)
+    {
+#if defined(ANDROID) || defined(IOS)
+      LogError("running a service node on mobile device is not possible.");
+      return false;
+#else
+      CURL *curl = curl_easy_init();
+      if(curl)
+      {
+        CURLcode res;
+        std::stringstream ss;
+        ss << "http://" << lokidRPCAddr << "/json_rpc";
+        const auto url = ss.str();
+        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+        curl_easy_setopt(curl, CURLOPT_HTTPAUTH, (long)CURLAUTH_ANY);
+        const auto auth = lokidRPCUser + ":" + lokidRPCPassword;
+        curl_easy_setopt(curl, CURLOPT_USERPWD, auth.c_str());
+        curl_slist *list = nullptr;
+        list = curl_slist_append(list, "Content-Type: application/json");
+        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, list);
+
+        nlohmann::json request = {{"id", "0"},
+                                  {"jsonrpc", "2.0"},
+                                  {"method", "get_service_node_privkey"}};
+        const auto data        = request.dump();
+        std::vector< char > resp;
+
+        curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, data.size());
+        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data.c_str());
+        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &resp);
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, &RecvIdentKey);
+        do
+        {
+          std::this_thread::sleep_for(std::chrono::milliseconds(100));
+          LogInfo("Getting Identity Keys from lokid...");
+          res = curl_easy_perform(curl);
+          if(res == CURLE_OK)
+          {
+            try
+            {
+              auto j = nlohmann::json::parse(resp);
+              if(not j.is_object())
+                continue;
+
+              const auto itr = j.find("result");
+              if(itr == j.end())
+                continue;
+              if(not itr->is_object())
+                continue;
+              const auto k =
+                  (*itr)["service_node_ed25519_privkey"].get< std::string >();
```

Comment:
You should `.find()` this and error out if it's not present in case someone runs it against a 5.x lokid.